ddprof-lib/src/main/cpp/objectSampler.cpp

```diff
@@ -64,25 +61,33 @@ void ObjectSampler::recordAllocation(jvmtiEnv* jvmti, JNIEnv* jni, jthread threa
         event._id = id;
     }
 
-    jvmtiFrameInfo *frames = new jvmtiFrameInfo[_max_stack_depth];
     jint frames_size = 0;
-    if (jvmti->GetStackTrace(thread, 0, _max_stack_depth,
-                                frames, &frames_size) != JVMTI_ERROR_NONE || frames_size <= 0) {
-        delete[] frames;
-        return;
-    }
-
+    jvmtiFrameInfo *frames = nullptr;
+    
+    // we do record the details and stacktraces only for when recording allocations or liveness
     if (_record_allocations || _record_liveness) {
-        std::set<jclass> classes;
-        jclass method_class;
-        for (int i = 0; i < frames_size; i++) {
-            if (jvmti->GetMethodDeclaringClass(frames[i].method, &method_class) == 0) {
-                classes.insert(method_class);
+        event._size = size;
+        event._weight =  (float)((size == 0 || _interval == 0) ? 1 : 1 / (1 - exp(-size / (double)_interval)));
+
+        frames = new jvmtiFrameInfo[_max_stack_depth];
```

Comment:
Can this use a linear allocator instead to avoid fragmentation?

Reply:
Linear allocator requires a clearly defined lifecycle. This is, unfortunately, not the case here.
I could try something like `_calltrace_buffer` but that would require locking. I did a quick check whether removing this allocation would help with the internal fragmentation (by having only one pre-allocated array and single lock, so reducing parallelism to 1) but the RSS growth I can observe is unaffected :(